File: 0907-koko-eating-bananas/0907-koko-eating-bananas.go
func minEatingSpeed(piles []int, h int) int {
    left , right :=1, max(piles)
    result := right
    
    for left<=right{
        mid := left +(right-left)/2
        if canFinish(piles, mid, h){
            result = mid 
            right = mid -1 
        }else{
            left = mid+1
        }
    }
    return result
}

func max(piles []int )int{
    maxVal:= 0
    for _, pile:= range piles{
        if pile > maxVal {
            maxVal = pile
        }
    }
    return maxVal
}
func canFinish(piles []int, k, h int) bool{
    hour:=0
    for _, pile:= range piles{
        hour += int(math.Ceil(float64(pile)/float64(k)))
    }
    return hour<=h 
}

File: 0907-koko-eating-bananas/0907-koko-eating-bananas.rs
impl Solution {
    pub fn min_eating_speed(piles: Vec<i32>, h: i32) -> i32 {
        let mut left = 1;
        let mut right = *piles.iter().max().unwrap();
        while left <right {
            let mid = left + (right- left)/2;
            if Self::can_eat_all(&piles, mid,  h){
                right = mid
            }else{
                left = mid+1;   
            }
        }
        left
    }
    fn can_eat_all(piles: &Vec<i32>, mid :i32,  h:i32)->bool{
        let hours: i32 = piles.iter()
                            .map(|&pile|(pile+mid-1)/mid)
                            .sum();

            hours<=h
    }
}